fish-rust/src/env/environment.rs

```diff
@@ -1,44 +1,462 @@
-// fish_read_limit variable.
-const DEFAULT_READ_BYTE_LIMIT: usize = 100 * 1024 * 1024;
-pub static mut read_byte_limit: usize = DEFAULT_READ_BYTE_LIMIT;
-pub static mut curses_initialized: bool = true;
+/// TODO: migrate to history once ported.
+const DFLT_FISH_HISTORY_SESSION_ID: &wstr = L!("fish");
 
+// Universal variables instance.
+lazy_static! {
+    static ref UVARS: Mutex<UniquePtr<env_universal_t>> = Mutex::new(env_universal_t::new_unique());
+}
+
+/// Set when a universal variable has been modified but not yet been written to disk via sync().
+static UVARS_LOCALLY_MODIFIED: RelaxedAtomicBool = RelaxedAtomicBool::new(false);
+
+/// Convert an EnvVar to an FFI env_var_t.
+fn env_var_to_ffi(var: EnvVar) -> cxx::UniquePtr<ffi::env_var_t> {
+    ffi::env_var_t::new_ffi(Box::into_raw(Box::from(var)).cast()).within_unique_ptr()
+}
+
+/// An environment is read-only access to variable values.
 pub trait Environment {
+    /// Get a variable by name using default flags.
     fn get(&self, name: &wstr) -> Option<EnvVar> {
-        todo!()
+        self.getf(name, EnvMode::DEFAULT)
     }
-    fn getf(&self, name: &wstr, mode: EnvMode) -> Option<EnvVar> {
-        todo!()
+
+    /// Get a variable by name using the specified flags.
+    fn getf(&self, name: &wstr, mode: EnvMode) -> Option<EnvVar>;
+
+    /// Return the list of variable names.
+    fn get_names(&self, flags: EnvMode) -> Vec<WString>;
+
+    /// Returns PWD with a terminating slash.
+    fn get_pwd_slash(&self) -> WString {
+        // Return "/" if PWD is missing.
+        // See https://github.com/fish-shell/fish-shell/issues/5080
+        let Some(var) = self.get(L!("PWD")) else {
+            return WString::from("/");
+        };
+        let mut pwd = WString::new();
+        if var.is_empty() {
+            pwd = var.as_string();
+        }
+        if !pwd.ends_with('/') {
+            pwd.push('/');
+        }
+        pwd
     }
+
+    /// Get a variable by name using default flags, unless it is empty.
     fn get_unless_empty(&self, name: &wstr) -> Option<EnvVar> {
-        todo!()
+        self.getf_unless_empty(name, EnvMode::DEFAULT)
     }
+
+    /// Get a variable by name using the given flags, unless it is empty.
     fn getf_unless_empty(&self, name: &wstr, mode: EnvMode) -> Option<EnvVar> {
-        todo!()
+        let var = self.getf(name, mode)?;
+        if !var.is_empty() {
+            return Some(var);
+        }
+        None
+    }
+}
+
+/// The null environment contains nothing.
+pub struct EnvNull;
+
+impl EnvNull {
+    pub fn new() -> EnvNull {
+        EnvNull
+    }
+}
+
+impl Environment for EnvNull {
+    fn getf(&self, _name: &wstr, _mode: EnvMode) -> Option<EnvVar> {
+        None
+    }
+
+    fn get_names(&self, _flags: EnvMode) -> Vec<WString> {
+        Vec::new()
+    }
+}
+
+/// A helper type for wrapping a type-erased Environment.
+pub struct EnvDyn {
```

Comment:
Haven't tried it but I'd remove this type and have functions take `&dyn Environment`, which is simpler.
(Of course we'd need to keep the FFI version)

Reply:
fixed